cpp/dolfinx/io/checkpointing.cpp

```diff
@@ -0,0 +1,151 @@
+// Copyright (C) 2024 Abdullah Mujahid
+//
+// This file is part of DOLFINX (https://www.fenicsproject.org)
+//
+// SPDX-License-Identifier:    LGPL-3.0-or-later
+
+#ifdef HAS_ADIOS2
+
+#include "checkpointing.h"
+#include "ADIOS2_utils.h"
+#include <adios2.h>
+#include <basix/finite-element.h>
+#include <dolfinx/mesh/Mesh.h>
+#include <mpi.h>
+
+using namespace dolfinx;
+using namespace dolfinx::io;
+
+/// @file checkpointing.h
+/// @brief ADIOS2 based checkpointing
+namespace
+{
+std::map<basix::element::lagrange_variant, std::string> lagrange_variants{
+    {basix::element::lagrange_variant::unset, "unset"},
+    {basix::element::lagrange_variant::equispaced, "equispaced"},
+    {basix::element::lagrange_variant::gll_warped, "gll_warped"},
+};
+
+template <std::floating_point T>
+void _write(ADIOS2Engine& adios2engine,
+            std::shared_ptr<dolfinx::mesh::Mesh<T>> mesh)
+{
+
+  auto io = adios2engine.io();
+  auto writer = adios2engine.engine();
+
+  const dolfinx::mesh::Geometry<T>& geometry = mesh->geometry();
+  auto topology = mesh->topology();
+
+  std::int16_t mesh_dim = geometry.dim();
```

Comment:
Is this performing a cast? Check definition.

Reply:
Yes it was, fixed.